hasjob/models/jobpost.py

```diff
@@ -225,66 +229,49 @@ def expiry_date(self):
-
-    def is_moderated(self):
-        return self.status == POSTSTATUS.MODERATED
-
-    def is_announcement(self):
-        return self.status == POSTSTATUS.ANNOUNCEMENT
-
-    def is_new(self):
-        return self.datetime >= datetime.utcnow() - newlimit
-
-    def is_closed(self):
-        return self.status == POSTSTATUS.CLOSED
-
-    def is_unacceptable(self):
-        return self.status in POSTSTATUS.UNACCEPTABLE
-
-    def is_old(self):
-        return self.datetime <= datetime.utcnow() - agelimit
-
-    def pay_type_label(self):
-        return PAY_TYPE.get(self.pay_type)
+    state.add_conditional_state('LISTED', state.PUBLIC, lambda jobpost: jobpost.datetime >= datetime.utcnow() - agelimit, label=('listed', __("Listed")))
+    state.add_conditional_state('CONFIRMABLE', state.UNPUBLISHED, lambda jobpost: jobpost.admin_is(current_auth.user), label=('confirmable', __("Confirmable")))
 
+    @state.transition(state.UNPUBLISHED, state.WITHDRAWN, title=__("Withdraw"), message=__("This job post has been withdrawn"), type='danger')
     def withdraw(self):
-        self.status = POSTSTATUS.WITHDRAWN
+        self.closed_datetime = db.func.utcnow()
 
+    @state.transition(state.PUBLIC, state.CLOSED, title=__("Close"), message=__("This job post has been closed"), type='danger')
     def close(self):
-        self.status = POSTSTATUS.CLOSED
+        self.closed_datetime = db.func.utcnow()
 
+    @state.transition(state.UNPUBLISHED, state.CONFIRMED, title=__("Confirm"), message=__("This job post has been confirmed"), type='success')
     def confirm(self):
-        self.status = POSTSTATUS.CONFIRMED
+        self.email_verified = True
+        self.datetime = db.func.utcnow()
+
+    @state.transition(state.PUBLIC, state.SPAM, title=__("Mark as spam"), message=__("This job post has been marked as spam"), type='danger')
+    def mark_spam(self, reason):
+        self.closed_datetime = db.func.utcnow()
+        self.review_datetime = db.func.utcnow()
+        self.review_comments = reason
+        self.reviewer = current_auth.user
+
+    @state.transition(state.DRAFT, state.PENDING, title=__("Mark as pending"), message=__("This job post is awaiting email verification"), type='danger')
+    def mark_pending(self):
+        pass
+
+    @state.transition(state.PUBLIC, state.REJECTED, title=__("Reject"), message=__("This job post has been rejected"), type='danger')
+    def reject(self, reason):
+        self.closed_datetime = db.func.utcnow()
+        self.review_datetime = db.func.utcnow()
+        self.review_comments = reason
+        self.reviewer = current_auth.user
+
+    @state.transition(state.PUBLIC, state.MODERATED, title=__("Moderate"), message=__("This job post has been moderated"), type='primary')
+    def moderate(self, reason):
+        self.closed_datetime = db.func.utcnow()
+        self.review_datetime = db.func.utcnow()
+        self.review_comments = reason
+        self.reviewer = current_auth.user
```

Comment:
Both these should accept `user` as a parameter. Don't assume `current_auth.user`.